new-frontiers-14/frontier-station-14
# Frontier Market / Trading console v0.1

## About the PR
This replaces https://github.com/new-frontiers-14/frontier-station-14/pull/1579.
The PR adds the Market console which is basically a buyback system. Its products will be entirely dependant upon players to sell to it by selling stuff through the currently available cargo selling stations. The market console will only hold prototype information so it will spawn new prototypes, erasing any data from whatever state it was sold as. Works with the market modifier component.
The data is stored on the grid it is sold on, and the 'cart' is kept on the console (for now).
Market data will stay as per-grid basis so you wont be able to buy from other stations afar, but we may expand this later on.

Because this system just receives events whenever something is sold from Cargo, it hooks into cargo events only and is a completely seperate system to minimize upstream merge conflicts.
Anything that can be sold through the regular cargo system is automatically recorded on the grid, and is filtered out on the server side when opening the market console. Because of this, cargo blacklists will automatically apply to it's contents.

TODO:
- [x] Deal with stacks (merging IronOre and IronOre1)
- [x] Add a blacklist (easy)
- [x] Fix crate machine animation bugs, it sometimes hangs so it doesnt deliver
- [x] transaction costs
- [x] enable/disable on power outage
- [x] case insensitive search
- [x] register container contents
- [x] Cart maximum contents of 30 items check
- [x] Limit purchases to maximum amount a generic steel crate can hold

Future plans:
- Contraband only market
- Checking the stock from handheld (pda?) device


## Why / Balance

This adds a new dimension to economy which we really need.

## How to test

This PR added the market console to cargo depots (for testing) that may get reverted.
- Teleport to a cargo depot using `tpto` command (use admin menus, last tab, to find the grid)
- Sell stuff
- Check market console on north side of the cargo depot
- Order stuff

## Media

![image](https://github.com/user-attachments/assets/34563ac9-22d6-4b77-b081-432e9dde3477)
![image](https://github.com/user-attachments/assets/07d9af11-9174-48a7-bebc-f7ca6ac91ff6)
![image](https://github.com/user-attachments/assets/f631ec8f-62ba-49c4-8467-e06112d60727)
![image](https://github.com/user-attachments/assets/7549a0f1-3606-4a95-bb0e-e42830154c45)




## Breaking changes


**Changelog**

:cl:
- add: Added market system
